                               RTF_REJECT patch
                               ----------------

RTF_REJECT.patch - apply this to 1.2.x Kernel with

   cd /usr/src
   patch -p0 -s < RTF_REJECT.patch

   cd linux ; make zLilo           # recompile Kernel

   This will add the RTF_REJECT route-flag as introduced in 1.3.x

   Please make sure to recompile the net-tools package! (route -V should
   show a +RTF_REJECT to use that feature!)

DESCRIPTION
-----------

RTF_REJECT - This allows to install a (net)'route' which will make the
   lookup for a route fail. This is used to notify the sender that the
   destination is unreachable. This is at the curent implementation not
   possible, if there is a default Gateway (which is very common on Systems
   with access to the Internet).

   RTF_REJECT is not primary for Firewalling (use IP_FIREWALL, ipfwadm
   instead) but to allow Linux to generate probably error Messages instead
   of Timeouts for unreachable (site-local) hosts.

   Example

-----------------------------------------
Kernel routing table
Destination     Gateway         Genmask         Flags MSS    Window Use Iface
localhost.      *               255.255.255.255 UH    1936   0        7 lo
host1.doma.in	*		255.255.255.255 UH    1436   0       10 eth0
host2.doma.in   *               255.255.255.255 UH    1436   0       12 eth0
doma-in-net     -               255.255.255.0   !     -      -        0 -
default         host1.doma.in   *               UG    1436   0      100 eth0
-----------------------------------------

   This Table is an example. localhost, host1 and host2 are set manually.
   doma-in-net is a blocking route installed with:

   route add -net doma-in-net reject

   In this example this particular hosts only routes packets to host1 host2
   of the domai-in-net. The World is still reachable. If the rejecting route
   is missing all acces to unknown hosts of the local net will be routed to
   the default gateway.

   An other Example:

-----------------------------------------
Kernel routing table
Destination     Gateway         Genmask         Flags MSS    Window Use Iface
localhost.      *               255.255.255.255 UH    1936   0        7 lo
host1.doma.in	-		255.255.255.255 !H    -      -       10 -
doma-in-net     *               255.255.255.0   U     1436   0        1 eth0
-----------------------------------------

   In this case a packet to host1 will be rejected with a 'now route to
   host' but all other hosts from the Class-C Network attached to the eth0
   can be accessed if present. If they are not present it ony will generate
   a timeout.


please give me Feedback about the Patches. They are intended to be
compatible to 1.3.x. In addition to 1.3.x Patches from Alan, my Patch also
counts the number of rejected Lookups.


bernd.eckenfels@inka.de

diff -u linux/net/inet/route.c.org linux/net/inet/route.c
--- linux/net/inet/route.c.org	Sun May  7 19:47:06 1995
+++ linux/net/inet/route.c	Mon Aug 14 02:26:10 1995
@@ -24,6 +24,7 @@
  *		Alan Cox	: 	MSS actually. Also added the window
  *					clamper.
  *		Sam Lantinga	:	Fixed route matching in rt_del()
+ *		Bernd Eckenfels :       RTF_REJECT Patch like 1.3.x Kernels
  *
  *		This program is free software; you can redistribute it and/or
  *		modify it under the terms of the GNU General Public License
@@ -562,6 +563,11 @@
 		if ((rt->rt_dev->flags & IFF_BROADCAST) &&
 		    (rt->rt_dev->pa_brdaddr == daddr))
 			break;
+	}
+	
+	if(rt->rt_flags&RTF_REJECT) {
+		rt->rt_use++;
+		return NULL;
 	}
 	
 	if(src_addr!=NULL)
diff -u linux/include/linux/route.h.org linux/include/linux/route.h
--- linux/include/linux/route.h.org	Sun May  7 19:39:08 1995
+++ linux/include/linux/route.h	Mon Aug 14 02:30:54 1995
@@ -58,6 +58,7 @@
 #define	RTF_MODIFIED	0x0020		/* modified dyn. (by redirect)	  */
 #define RTF_MSS		0x0040		/* specific MSS for this route	  */
 #define RTF_WINDOW	0x0080		/* per route window clamping	  */
+#define RTF_REJECT	0x0200		/* Reject route			  */
 
 /*
  *	REMOVE THESE BY 1.2.0 !!!!!!!!!!!!!!!!!
diff -u linux/net/inet/route.h.org linux/net/inet/route.h
--- linux/net/inet/route.h.org	Tue Aug 15 00:07:09 1995
+++ linux/net/inet/route.h	Tue Aug 15 00:07:09 1995
@@ -12,6 +12,7 @@
  * Fixes:
  *		Alan Cox	:	Reformatted. Added ip_rt_local()
  *		Alan Cox	:	Support for TCP parameters.
+ *		Bernd Eckenfels :	rt_flag should be short for 1.3
  *
  *		This program is free software; you can redistribute it and/or
  *		modify it under the terms of the GNU General Public License
@@ -32,7 +33,7 @@
 	unsigned long		rt_dst;
 	unsigned long		rt_mask;
 	unsigned long		rt_gateway;
-	unsigned char		rt_flags;
+	unsigned short		rt_flags;
 	unsigned char		rt_metric;
 	short			rt_refcnt;
 	unsigned long		rt_use;
